cataclysmbnteam/Cataclysm-BN
# [Rdy?]Redo random NPC spawn chance

#### Summary



SUMMARY: Balance "Redo random NPC spawn chance"


#### Purpose of change



Make random NPCs actually spawn in reasonable numbers.

#### Describe the solution



* Changed the NPC search radius and spawn radius the same (120 tiles, was 60 for search before)
* Changed the spawn chance to:
 * `density / 24` per hour until it reaches (very) soft cap
 * `(1 / 24) / (npc_count / cap)` above the cap
* Cap is number of days in a year (`4 * 14`, of course) multiplied by density
* The result is NPC count rising with `sqrt(t)`, where `t` is time, assuming none die and the player is stationary
* Prevented random NPCs spawning within 30 overmap tiles of player
* Gave NPCs 90 extra chances to pick a location (was 10) that isn't river or lake, and now also within 30 tiles of player

#### Describe alternatives you've considered



* No penalty for overcrowding
* Reworking random NPC spawns, not just the chance, and having another unfinished PR sit in the backlog
* Giving random NPCs a tag, counting only those with the tag when calculating the chance, to skip static NPCs

#### Testing



* Enable debug logging in options
* Enable random NPCs and set random NPC multiplier to 100
* Wait for 300 minutes
* Open `config/debug.log`
* Note 5-6 lines of "spawning random NPC at"
* Set the NPC multiplier to 1
* Enable debug mode (the spammy one)
* Wait to exact point hour changes (so 7:59 to 8:00 etc.)
* Notice the line about spawn chance % and that it looks like some sane percentage

#### Additional context



I made plots of generated NPC count vs hours and vs sqrt of hours.
I used brute force instead of proper math, so it's wobbly. Each line is a separate simulation, but you can see they're very close.
The second plot is for square root of hours. As you can see, it becomes pretty much linear later.

![npcgen](https://user-images.githubusercontent.com/1878860/237005999-7681bbdd-968a-484c-9be2-8d4879369ef8.png)
![npcgen-sqrt](https://user-images.githubusercontent.com/1878860/237006005-3f040dbb-7022-44e7-a1b2-2d563b1a205e.png)
